storage_controller/src/service.rs

```diff
@@ -392,14 +392,41 @@ pub struct Config {
     /// tenant-scoped API endpoints. Further API requests queue until ready.
     pub tenant_rate_limit: NonZeroU32,
 
-    /// The size at which an unsharded tenant should be split (into 8 shards). This uses the logical
-    /// size of the largest timeline in the shard (i.e. max_logical_size).
+    /// If a tenant shard's largest timeline (max_logical_size) exceeds this value, all tenant
+    /// shards will be split in 2 until they fall below split_threshold (up to max_split_shards).
+    ///
+    /// This will greedily split into as many shards as necessary to fall below split_threshold, as
+    /// powers of 2: if a tenant shard is 7 times larger than split_threshold, it will split into 8
+    /// immediately, rather than first 2 then 4 then 8.
     ///
     /// None or 0 disables auto-splitting.
     ///
     /// TODO: consider using total logical size of all timelines instead.
     pub split_threshold: Option<u64>,
 
+    /// The maximum number of shards a tenant can be split into during autosplits. Does not affect
+    /// manual split requests. 0 or 1 disables autosplits.
+    pub max_split_shards: u8,
```

Comment:
nit: prefer Option to numbers for disabling (magic numbers at CLIs are okay, but if possible transform into Option for internal use)

Reply:
although reading further... I kind of see why you did it this way.  0 and 1 aren't really magic numbers, it's just that the consequence of saying "at most 0/1 shards" is that you don't get multi-sharded tenants, so it's more of a silly but handleable input. 

never mind